pandas/_libs/src/vendored/ujson/python/objToJSON.c

```diff
@@ -492,7 +490,8 @@ static int NpyArr_iterNextItem(JSOBJ obj, JSONTypeContext *tc) {
     ((PyObjectEncoder *)tc->encoder)->npyValue = npyarr->dataptr;
     ((PyObjectEncoder *)tc->encoder)->npyCtxtPassthru = npyarr;
   } else {
-    GET_TC(tc)->itemValue = npyarr->getitem(npyarr->dataptr, npyarr->array);
+    GET_TC(tc)->itemValue =
+        PyArray_GETITEM((PyArrayObject *)npyarr->array, npyarr->dataptr);
```

Comment:
Would still prefer a type check before the cast - prevents strange bugs in case this ever gets refactored.

Reply:
Looks like it already happens a couple lines above.
https://github.com/pandas-dev/pandas/blob/e14a9bd41d8cd8ac52c5c958b735623fe0eae064/pandas/_libs/src/vendored/ujson/python/objToJSON.c#L477-L480

Reply:
In that case we should assign `const PyArrayObject *arrayobj = (const PyArrayObject *)npyarr->array` right after the type check and reference that here.

Generally want to minimize casting and keep it very localized to where the type check happens (historically we have not done a good job of this) Otherwise things are very liable to break when refactoring